src/Storages/ExternalStream/Kafka/KafkaSink.h

```diff
@@ -56,42 +52,62 @@ class KafkaSink final : public SinkToStorage
     void checkpoint(CheckpointContextPtr) override;
 
 private:
-    static void onMessageDelivery(rd_kafka_t * /* producer */, const rd_kafka_message_t * msg, void * opaque)
-    {
-        static_cast<KafkaSink *>(opaque)->wb->onMessageDelivery(msg);
-    }
-
-    static int32_t onPartitioning(
-        const rd_kafka_topic_t * /*rkt*/,
-        const void * /*keydata*/,
-        size_t /*keylen*/,
-        int32_t partition_count,
-        void * rkt_opaque,
-        void * msg_opaque)
-    {
-        /// update partition count
-        auto * sink = static_cast<KafkaSink *>(rkt_opaque);
-        sink->partition_cnt = partition_count;
-
-        auto parition_id = static_cast<Int32>(reinterpret_cast<std::uintptr_t>(msg_opaque));
-        /// This should not really happen because Kafka does not support reducing partitions.
-        /// However, KIP-694 is currently under discussion, so this might heppen in the future.
-        if (parition_id >= partition_count)
-            parition_id = partition_count - 1;
-        return parition_id;
-    }
+    /// Callback for Kafka message delivery report
+    static void onMessageDelivery(rd_kafka_t * /* producer */, const rd_kafka_message_t * msg, void *  /*opaque*/);
+    void onMessageDelivery(const rd_kafka_message_t * msg);
+
+    void addMessageToBatch(char * pos, size_t len);
+
+    /// the number of acknowledgement has been received so far for the current checkpoint period
+    size_t acked() const { return state.acked; }
+    /// the number of errors has been received so far for the current checkpoint period
+    size_t error_count() const { return state.error_count; }
```

Comment:
`errorCount()` and i usually put `noexcept` there for most setter / getter if they are no-memory allocation and we are sure it won't throw. like `errorCount() const noexcept`